mirascope/core/base/_utils/_convert_messages_to_message_params.py

```diff
@@ -0,0 +1,92 @@
+from __future__ import annotations
+
+from collections.abc import Sequence
+from typing import TYPE_CHECKING
+
+from typing_extensions import TypeIs
+
+from .. import AudioPart, BaseMessageParam, CacheControlPart, ImagePart, TextPart
+from .._utils import get_image_type
+
+if TYPE_CHECKING:
+    from PIL import Image
+
+    has_pil_module: bool
+else:
+    try:
+        from PIL import Image  # pyright: ignore [reportAssignmentType]
+
+        has_pil_module = True
+    except ImportError:  # pragma: no cover
+        has_pil_module = False
+
+        class Image:
+            class Image:
+                def tobytes(self) -> bytes: ...
+
+
+def _convert_message_sequence_part_to_content_part(
+    message_sequence_part: str | Image.Image | TextPart | ImagePart | AudioPart,
+) -> TextPart | ImagePart | AudioPart | CacheControlPart:
+    if isinstance(message_sequence_part, str):
+        return TextPart(text=message_sequence_part, type="text")
+    elif isinstance(
+        message_sequence_part, TextPart | ImagePart | AudioPart | CacheControlPart
+    ):
+        return message_sequence_part
+    elif has_pil_module and isinstance(message_sequence_part, Image.Image):
+        image = message_sequence_part.tobytes()
+        return ImagePart(
+            type="image",
+            media_type=f"image/{get_image_type(image)}",
+            image=image,
+            detail=None,
+        )
+    else:
+        raise ValueError(f"Invalid message sequence type: {message_sequence_part}")
+
+
+def get_content_from_message(
+    content: str | Sequence[str | Image.Image | TextPart | ImagePart | AudioPart],
+) -> str | list[TextPart | ImagePart | AudioPart | CacheControlPart]:
+    if isinstance(content, str):
+        return content
+    return _convert_message_sequence_to_content(content)
+
+
+def _convert_message_sequence_to_content(
+    message_sequence: Sequence[str | Image.Image | TextPart | ImagePart | AudioPart],
+) -> list[TextPart | ImagePart | AudioPart | CacheControlPart]:
+    return [
+        _convert_message_sequence_part_to_content_part(message)
+        for message in message_sequence
+    ]
+
+
+def _is_base_message_params(
+    value: object,
+) -> TypeIs[list[BaseMessageParam]]:
+    return isinstance(value, list) and isinstance(value[0], BaseMessageParam)
+
+
+def convert_messages_to_message_params(
+    messages: str
+    | Sequence[str | Image.Image | TextPart | ImagePart | AudioPart]
+    | list[BaseMessageParam]
+    | BaseMessageParam,
+    role: str = "user",
+) -> list[BaseMessageParam]:
+    if isinstance(messages, str):
+        return [BaseMessageParam(content=messages, role=role)]
+    elif isinstance(messages, BaseMessageParam):
+        return [messages]
+    elif _is_base_message_params(messages):
+        return messages
+    elif isinstance(messages, Sequence):
+        return [
+            BaseMessageParam(
+                content=_convert_message_sequence_to_content(messages), role=role
```

Comment:
wrt. comment before, content here can be a `str` type so no need for a separate method, we can just use the merged `convert_message_content_to_message_param_content`